lib/Corres_Cases.thy

```diff
@@ -0,0 +1,319 @@
+  by (clarsimp simp add: corres_case_helper_def subset_iff)
+
+private lemma corres_case_imp_process_weak:
+  "\<lbrakk> S \<Longrightarrow> corres_case_helper (P, P') (Q, Q') R \<rbrakk>
+   \<Longrightarrow> corres_case_helper (P, P') (\<lambda>s. S \<longrightarrow> Q s, Q') (S \<longrightarrow> R)"
+  by (clarsimp simp add: corres_case_helper_def subset_iff)
+
+private lemmas corres_case_process =
+  corres_case_conj_process corres_case_all_process corres_case_imp_process
+
+private lemmas corres_case_process_weak =
+  corres_case_conj_process corres_case_all_process_weak corres_case_imp_process_weak
+
+(* Turn goals of the form
+
+   (\<forall>y. x = SomeConstr y \<longrightarrow> corres (?P x) P' (SomeConstr y) g) \<and>
+   (\<forall>y. x = OtherConstr y \<longrightarrow> corres (?P x) P' (SomeConstr y) g) \<and>
+   ...
+
+   into multiple goals of the form
+
+   \<And>y. x = SomeConstr y \<Longrightarrow> corres (?P1 x y) ?P'1 (SomeConstr y) g)
+   \<And>y. x = OtherConstr y \<Longrightarrow> corres (?P2 x y) ?P'2 (SomeConstr y) g)
+
+   with instantiations
+
+   ?P x = \<lambda>s. (\<forall>y. x = SomeConstr y \<longrightarrow> ?P1 x y s) \<and> (\<forall>y. x = OtherConstr y \<longrightarrow> ?P2 x y s)
+   ?P' = \<lambda>s. ?P'1 s \<and> ?P'2 s
+
+   We do this by first transforming the goal into a corres_case_helper goal, and then applying
+   the corresponding lifting rules. We first try if we can get both sides (?P and ?P') to have
+   quantifiers and implications to get a stronger statement, and fall back to the weaker \<and> for ?P'
+   shown above when that doesn't work (e.g. because ?P' might not depend on x).
+
+   When all lifting rules have applied, we transform the goal back into a corres goal using the
+   provided helper rule (e.g. corres_case_helper_corres_left below).
+*)
+private method corres_cases_body uses helper =
+  determ \<open>rule corres_case_helperI, repeat_new \<open>rule corres_case_process\<close>; rule helper
+          | rule corres_case_helperI, repeat_new \<open>rule corres_case_process_weak\<close>; rule helper\<close>
+
+
+(* Instances of corres_case_helper for left and right side of the corres predicate.
+   These lemmas bind the corres guards to the corres_case_helper guards. *)
+private lemma corres_case_helper_corres_left:
+  "corres_underlying sr nf nf' r Q Q' f f' \<Longrightarrow>
+    corres_case_helper (P, P') (Q, Q') (corres_underlying sr nf nf' r P P' f f')"
+  by (auto simp: corres_case_helper_def elim!: corres_guard_imp)
+
+private lemma corres_case_helper_corres_right:
+  "corres_underlying sr nf nf' r Q' Q f f' \<Longrightarrow>
+    corres_case_helper (P, P') (Q, Q') (corres_underlying sr nf nf' r P' P f f')"
+  by (auto simp: corres_case_helper_def elim!: corres_guard_imp)
+
+
+section \<open>Main method definitions\<close>
+
+(* Case distinction on abstract side *)
+method corres_cases_left =
+  determ \<open>
+    corres_pre,
+    (match conclusion in
```

Comment:
how are you managing to do an Eisbach `match` with schematics around? I thought `match` forced variable focus

Reply:
Interesting, I adapted that from Dan's code without thinking about schematics. You're right that we'd usually expect them to be frozen. I *think* what is going on is that the pattern matches the schematics explicitly with `for .. P P' ..` which then makes them schematic again on the inside. All other schematics would remain frozen, but that is even  beneficial in this context, because we would not want to touch those.

Reply:
My concern is that "freezing" a schematic *is* touching it, i.e. you cannot instantiate it after it is frozen, which is why we usually can't use `match` in wp proofs. This would indicate that there is a workaround available, but it will only work if the match can mention and re-generalise the schematics directly? Something to think about for sure.

If we end up with a schematic in the assumptions, probably that's a mistake. We really shouldn't see any schematics in the corres bodies... so that leaves the return relation. I don't immediately know of any cases where we'd want to do `corres_cases` with a schematic return relation. What do you think?

Reply:
The freezing is only inside the scope of the `match`, and once the `match` is complete they unfreeze, so if we don't need to interact with them, they should be unchanged. That means schematic return relations should be fine.

For implementing `wp` that is not useful enough, because there we'd usually want them to stay schematic on the inside of the `match` scope, but you are right that explicitly matching them in patterns like `for .. P P' ..` might be a workaround for that. Definitely worth keeping in mind.

Reply:
On second reading, there is unfortunately nothing new here at all: the `for .. P P' ..` doesn't make the `P` schematic inside the `match`, they just become fixed variables. What's going on here is that the `match` scope is small, it only covers `apply_split`, which figures out and the split rule, but does not touch any schematics. The precondition processing happens after all that, when the `match` statement is concluded and the schematics are available again.

Reply:
It seems I do not understand variable focus then. My impression was if you focus, *all* schematics freeze (become fixed variables), and then you can't apply schematic manipulation functions as in `wp` goals. I was not aware that if you apply a rule that does not affect the schematic, after the entire application (match + rule) is applied, the schematics unfreeze ... is that right?

Reply:
> It seems I do not understand variable focus then. My impression was if you focus, _all_ schematics freeze (become fixed variables), and then you can't apply schematic manipulation functions as in `wp` goals. 

That's exactly right.

> I was not aware that if you apply a rule that does not affect the schematic, after the entire application (match + rule) is applied, the schematics unfreeze ... is that right?

The schematics always unfreeze after the match, not matter what you do. The effect is that inside the match you can't instantiate schematics that existed when you matched, but you can do everything else. The main thing that it is trying to make sure is that the focussed proof does not affect state outside the focussed proof (which instantiated schematics would).

Reply:
OK, that makes `match` less useless in schematic goals than I thought! It means that you could maybe use them to apply a rule to a `wp` goal to modify the head of the bind, for example, so long as it can deal with the schematics being frozen. That opens up a new avenue of thinking that I didn't think was possible, thank you so much for explaining.